### ISSUE TITLE
vidyut-lipi needs to handle the colon separator in ISO-15919

### ISSUE THREAD

**Issue by deepestblue** [ambuda-org/vidyut#103]
One of the many corner cases ISO-15919 supports is using a `:` to disambiguate Latin letter clusters. Here're a couple examples that need support in vidyut-lipi.

`./lipi -f devanagari -t iso19519 "अर्शइत्यादयः"`

Expected: `arśa:ityādayaḥ`
Actual: `arśaityādayaḥ`

./lipi -t devanagari -f iso19519 "arśa:ityādayaḥ"
Expected: `अर्शइत्यादयः`
Actual: `अर्श:इत्यादयः`

`./lipi -f devanagari -t iso19519 "वाग्हरि"`
Expected: `vāg:hari`
Actual: `vāghari`

`./lipi -t devanagari -f iso19519 "vāg:hari"`
Expected: `वाग्हरि`
Actual: `वाग्:हरि`

**Comment by deepestblue** [comment]
SaulabhyaJS also supports it, if you want to take a look. https://github.com/deepestblue/saulabhyaJS/blob/main/src/saulabhya.js and search for `separator`

**Comment by deepestblue** [comment]
1 and 3 sound right to me. On item 2, given Sanskrit doesn't use Latin punctuation traditionally and even in modern Sanskrit, people generally only use the comma, the question mark and the exclamation mark (because I guess the colon is rare even in English), I'd maybe propose instead to error out if Latin input contains the colon other than in these specified contexts?

**Comment by deepestblue** [comment]
Hmm, my 2 cents is that I'd expect a transliterator like this to be very conservative on input-handling; otherwise round-tripping gets messy, behaviour becomes fuzzy, etc.

I'd propose that un-proofread content isn't a valid scenario.

As for mixed-content, my thought is that the content could be marked up appropriately outside of invoking this library. Say in HTML, the markup can contain the `lang` attribute, and the JS that would invoke vidyut would invoke it only for the appropriately marked up nodes.

**Comment by shreevatsa** [comment]
Possible examples of the options I mean:

* for the case from the other bug, an option for whether "rR" should be transliterated into [‎0930 DEVANAGARI LETTER RA,  ‎094D DEVANAGARI SIGN VIRAMA,  ‎090B DEVANAGARI LETTER VOCALIC R] or into [‎0930 DEVANAGARI LETTER RA, ‎0943 DEVANAGARI VOWEL SIGN VOCALIC R]. (Rendered the same in some fonts?)
* What to do with "कइ" (transliterate as "kai", "ka{}i", "käi", return two separate chunks for the wrapper to deal with, …).
* Whether to treat a colon as visarga, pass it through as a colon, or throw an error. (A `colon_strategy` field of the options struct parameter?)
* How to deal with short e/o when encountered in non-Devanagari (transliterate to Devanagari short e/o that are strictly correct but many people don't recognize and possibly some fonts too, or the regular long ones?), or in Devanagari (some people seem to be using text input methods that produce these characters even when they clearly intend the regular long ones).
* Many more (see Aksharamukha UI for a few: replace anusvāra with corresponding anunāsika or vice-versa, etc)…

Even if we expect very few people to use the transliterator "core" function directly, it would be a way of writing down explicitly all the choices that have been made in the convenience wrapper.

**Comment by shreevatsa** [comment]
Ha, I missed that this discussion was about treating colon as a separator, which is relevant to two of my examples above :)

Also more concretely responding to comment https://github.com/ambuda-org/vidyut/issues/103#issuecomment-1909404083 above, rather than 

```
struct Result {
  text: String
  errors: Vec<ErrorSpan>
}

struct ErrorSpan {
  // Byte offsets in the input string. `usize` = platform-specific unsigned int, e.g. u64
  start: usize
  end: usize
  error: ...
}
```

where the consumer has to manually match up the best-effort text with byte offsets, what I was proposing is something like (may not be working code, treat as pseudocode):

```
// result: Vec<ResultChunk>

struct ResultChunk {
    text: String,
    kind: ResultKind,
}

enum ResultKind {
    Fine(String), // perfectly fine and unproblematic input for the source and destination scripts: well-understood and will round-trip cleanly
    UnknownPassedThrough(String), // emoji, punctuation, etc: not part of the source and destination scripts, but just passed through
    LikelyInputErrorSilentlyCorrected(String), // e.g. "s" in Devanagari corrected to avagraha
   // ...
}
```

or whatever, and the default convenience wrapper would just concatenate all the result chunks' `text` while the "serious" user could assemble their own different result by looking into the `ResultKind`s.

**Comment by shreevatsa** [comment]
Transliterating from a script to itself (Devanagari to Devanagari, or IAST to IAST) would then be a way of finding all problematic stuff in it :-)

Anyway I'll stop the flood of comments here; aware that what I'm proposing is likely overengineering :-) The broader point is just a desire for a really conservative/pedantic/lossless transliterator which will never silently corrupt text no matter what the input is or how many rounds of transliteration in whatever directions it undergoes using the library.

**Comment by deepestblue** [comment]
A couple of sorta related issues

`aū` should transliterate  to `अऊ`

`agḥ` should transliterate to `अग्ः` (I'm not sure there's a use-case for this specific example)

**Comment by deepestblue** [comment]
Yep, this seems similar to the code in saulabhyaJS near https://github.com/deepestblue/saulabhyaJS/blob/main/src/saulabhya.js#L352
